docs/developers/source-code.rst

```diff
@@ -0,0 +1,265 @@
+  * The one exception to this is that when doing an assignment to a
+    shell variable from another shell variable, it is not necessary to
+    use quotes on the right hand side:
+
+    .. code-block:: sh
+
+       # This is harmless, but unnecessary
+       foo="$bar"
+
+       # This is actually sufficient, even for multi-token values of $bar
+       foo=$bar
+
+* Do not use the ``==`` operator for ``test`` |mdash| this is a GNU
+  extension and can cause portability problems on BSD systems.
+  Instead, use the single ``=`` operator.
+
+  .. code-block:: sh
+
+     # This is bad
+     if test "$foo" == "bar"; then
+
+     # This is good
+     if test "$foo" = "bar"; then
+
+* Do not use the ``-a`` or ``-o`` operators for ``test`` |mdash| this
+  has caused portability problems with ``test(1)`` on BSD systems.
+  Instead, use the ``&&`` or ``||`` shell operators.
+
+  .. code-block:: sh
+
+     # This is bad
+     if test "$foo" = "bar" -a "$baz" = "yow"; then
+
+     # This is good
+     if test "$foo" = "bar" && test "$baz" = "yow"; then
+
+m4
+^^^
+
+We do not have specific coding style guidelines for m4 (the language
+used to create the ``configure`` script).  Please read some of the
+existing m4 code in the source code tree and try to use a similar
+style.
+
+Tree layout
+-----------
+
+There are a few notable top-level directories in the source
+tree:
+
+* The main sub-projects:
+
+    * ``oshmem``: Top-level OpenSHMEM code base
+    * ``ompi``: The Open MPI code base
+    * ``opal``: The OPAL code base
+
+* ``config``: M4 scripts supporting the top-level ``configure`` script
+  ``mpi.h``
+* ``etc``: Some miscellaneous text files
+* ``docs``: Source code for Open MPI documentation
+* ``examples``: Trivial MPI / OpenSHMEM example programs
+* ``3rd-party``: Included copies of required core libraries (either
+  via Git submodules in Git clones or via binary tarballs).
+
+  .. note:: While it may be considered unusual, we include binary
+     tarballs (instead of Git submodules) for 3rd party projects that
+     are:
+
+     #. Needed by Open MPI for correct operation, and
+     #. Not universally included in OS distributions, and
+     #. Rarely updated.
+
+Each of the three main source directories (``oshmem``, ``ompi``, and
+``opal``) generate at least a top-level library named ``liboshmem``,
+``libmpi``, and ``libopen-pal``, respectively.  They can be built as
+either static or shared libraries.  Executables are also produced in
+subdirectories of some of the trees.
+
+Each of the sub-project source directories have similar (but not
+identical) directory structures under them:
+
+* ``class``: C++-like "classes" (using the OPAL class system)
+  specific to this project
+* ``include``: Top-level include files specific to this project
+* ``mca``: MCA frameworks and components specific to this project
+* ``runtime``: Startup and shutdown of this project at runtime
+* ``tools``: Executables specific to this project (currently none in
+  OPAL)
```

Comment:
We have the opal wrapper compilers down there. So it's a bit fuzzy to say that there are none in OPAL. Probably best to remove the note.
```suggestion
* ``tools``: Executables specific to this project
```

Reply:
This is actually old text -- I didn't even edit anything in this area.  But this is a good update -- I'll add it.